[Mobile] Restyle product modal

luisramos0 opened the pull request:
#### What? Why?

Closes #4684 

Restyle product modal.
Checklist from issue:

- [x] Modals should present a clear 'outside area' where the user can tap/click to close the modal (see designs for mobile / tablet / desktop padding)
- [x] As per current production, Producer Name IS NOT a clickable link so users don't let lost in a myriad of modals
- [x] Remove product category
- [x] Display product property(s) in the smaller size as per new shop product listing
- [x] For modals which require scrolling (uncommon but possible, esp. on mobile) display a clear 'close' button at the bottom of the modal
- [x] Remove grey border around images

I have not managed to do this one in a reliable way, unable to override zerb foundations values. We can create a separate issue for it:
- [ ] Modals should be centered in the viewport (currently in mobile they're squished to the top)

The "outside area' change was done on all frontoffice modals as discussed in the issue.
So the following modals were also affected and will need to be tested: login, registration and cookies policy.

Some screenshots:
![image](https://user-images.githubusercontent.com/1640378/80619246-5b245500-8a3c-11ea-88ea-06c8336138ef.png)

![image](https://user-images.githubusercontent.com/1640378/80523571-3077c500-8986-11ea-8665-7f8034960eef.png)

![image](https://user-images.githubusercontent.com/1640378/80619349-8313b880-8a3c-11ea-873f-c49bb3a07b46.png)

![image](https://user-images.githubusercontent.com/1640378/80619515-b3f3ed80-8a3c-11ea-95a5-30ffa354f110.png)

![image](https://user-images.githubusercontent.com/1640378/80619306-755e3300-8a3c-11ea-8ec5-e8a3457b56ae.png)



Login:
![image](https://user-images.githubusercontent.com/1640378/80619575-c9691780-8a3c-11ea-98db-b6c5a56960f7.png)

Registration:
![image](https://user-images.githubusercontent.com/1640378/80523318-c6f7b680-8985-11ea-917d-fcce84396e74.png)

Cookies:
![image](https://user-images.githubusercontent.com/1640378/80523466-10e09c80-8986-11ea-96de-fda5f979cbef.png)


#### What should we test?
We need to test the changes described above on the product modal.
On the extra modals (login, registration process and cookies policy) we just need to verify that the margin of the modal looks good.

#### Release notes
<!-- Write a line or two to be included in the release notes.
Everything is worth mentioning, because you did it for a reason. -->



<!-- Please assign one category to your PR and delete the others. 
The categories are based on https://keepachangelog.com/en/1.0.0/. -->

Changelog Category: Added | Changed | Deprecated | Removed | Fixed | Security

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-modals.css.scss:
```diff
@@ -1,9 +1,30 @@
+    a {
+      border: 1px solid $grey-300;
+      font-weight: normal;
+      padding: 0.2em 0.625em;
```
what not `rem` here? won't make it harder to change if we don't stick to `rem` as we do in the rest of the file?

luisramos0 replied:
I think I copied this from the shop filters. I changed it now to rem.

sauloperez commented on app/assets/stylesheets/darkswarm/images.css.scss:
```diff
@@ -17,6 +14,12 @@
+
+  @media only screen and (max-width: 1024px) {
+    margin-top: 0;
+    margin-bottom: 1em;
```
I think we should find consensus on whether we use `rem` or `em` in our designs, don't you think?

luisramos0 replied:
well, it's a design decision isnt it? do you want the element to scale with the font size or not. I think in most cases you do, so REM should be used virtually everywhere.

I had a look at this one, and I just removed this margin, it's not necessary.

sauloperez commented on app/assets/stylesheets/darkswarm/modals.css.scss:
```diff
@@ -72,6 +77,20 @@ dialog .mobile-close-reveal-modal
+    background-color: $orange-500;
+    margin: 0 0 1rem;
+
+    @media only screen and (min-width: 640px) {
```
haven't we merged @Matt-Yorkley 's mixin to abstract media breakpoints?

Matt-Yorkley replied:
Nope, those changes are in the feature branch which is currently blocked by: #5345

luisramos0 replied:
I cant use it yet, not in master...

sauloperez commented:
:heart_eyes: I love how this is looking!

sauloperez commented on app/assets/stylesheets/darkswarm/modals.css.scss:
```diff
@@ -18,23 +16,26 @@ dialog
-  // Medium and up - when modal IS NOT full screen
+  // Medium and up - larger outside area
   @media only screen and (min-width: 641px) {
+    border-bottom: 30px solid white;
```
don't we have a var for `white`? if we ever decide we go with a slightly off-white (very common) we'll have to go replace all uses of `white`.

sauloperez replied:
I see we already do that with `$black` :point_down:

Matt-Yorkley replied:
We do have $white already.

Matt-Yorkley replied:
We do have $white already.

luisramos0 replied:
:+1:

sauloperez commented on engines/web/app/assets/stylesheets/web/pages/cookies_policy_modal.css.scss:
```diff
@@ -4,6 +4,10 @@
   border-bottom-color: $cookies-policy-modal-border-bottom-color;
 
+  @media only screen and (max-width: 640px) {
+    padding: 15px 15px 0 15px;
```
I've seen those 15px many times already. I'd be great if we could abstract it somehow. What about $gap, `$gap-sm` or something alike? we need to watch after consistency and the only way to enforce it is by sticking to a style-guide sort of vars.

Matt-Yorkley replied:
1rem is 16px right? Maybe we could use that in place if 15px here?

sauloperez replied:
assuming the font-size we set to the HTML tag is 16px, yes. And yes, that sounds more consistent.

luisramos0 replied:
:+1: changed to rem

luisramos0 commented:
Moving to in dev until dev (me) addreses review feedback

luisramos0 commented:
rebased to resolved conflict

luisramos0 commented:
@sauloperez I added one commit to use rem in some places and use the $white variable :+1:

yukoosawa commented:
Testing this on AU staging (with iPhone 6S and iPad 3 mini) and there are some issues :(

The background scroll behind the open modal in **mobile/tablet** is causing issues -- however this problem exists in production (booo!). @daniellemoorhead will create a bug ticket to address separately #5397 which will block this ticket for now. Issues include:

- User can scroll the modal completely out of the viewport (and then get stuck and not be able to scroll back up the page)
- Modals can be opened and positioned off the viewport (login/signin)
- 'Close' button appears in modals that don't require it post background scroll

A few other things to address:

1. Can we add padding to the bottom of the image. In mobile/tablet the close button is too close to the image.

![IMG_0943](https://user-images.githubusercontent.com/19785399/81488469-30ee4680-9237-11ea-9b56-4d27a68ea490.png)

2. The close button scrolling away doesn't seem right - but recognise the button's current position can't be fixed as the content will collide with it when scrolled. @luisramos0 What do you think about repositioning it partially off the modal? This would remove the need to have the 'close' button at the bottom of the modal.

![Mobile: Product Detail Modal](https://user-images.githubusercontent.com/19785399/81489010-b3c6cf80-923e-11ea-8466-d5a5881463af.png)

luisramos0 commented:
Thanks a lot for testing.
Regarding the first point: we dont usually block PRs because of issues that already exist in production.
There's a cost to keeping PRs in progress (code conflicts, repeat code review, etc).
We usually say the PR improves the situation in prod so we ship it and get the remaining issues sorted in separate PRs.
Is there a strong reason for not shipping this without 5397?  

I am going to address the last 2 points of your comment in this PR tomorrow :+1:

luisramos0 commented:
ok, I have addressed point 1 above, padding below image: my rectangular test image was not a good image for testing, the square image is better :+1: I think it's good now with this last commit :+1:
I just changed the values of two paddings so I'd say it doesnt need code review: this PR can be re-staged and re-tested.

Regarding point 2, the close button is scrolling away in production. And that was not required or changed in this issue/PR so I'd say we should move this PR forward as is and address that in a separate issue/PR.
I am saying that to avoid confusion but also because it feels it still needs a design decision?
I think this discussion should be in an issue not in a PR but here are my opinions:
-  if we make the top X close button fixed, the close button at the bottom added in this PR can be removed?
- regarding the other option, to move the top X button outside the modal: I think it will be tricky to keep the button always exactly in that position...

I'd say that, to keep it simple, we should ship this PR as is (we need to test this last correction, point 1) and address the two remaining issues, both already in production, in separate issues/PRs.

sauloperez commented:
> regarding the other option, to move the top X button outside the modal: I think it will be tricky to keep the button always exactly in that position...

a bit of negative top and right margin should pull the X outside of its parent. If not, there's the option of using `position: absolute` and negative `top` and `right`.

luisramos0 commented:
@sauloperez yeah, I think I have seen the button float a little with different screen sizes. The other issue is: how do I make it show above the modal background? (z-index is not helping)
![image](https://user-images.githubusercontent.com/1640378/81839492-83817880-953f-11ea-8dda-a6f2140d856e.png)

daniellemoorhead commented:
I'm going to focus on this today, clear the block and make a decision on how to move this and the bug forward. Stay tuned.

daniellemoorhead commented:
Hey @luisramos0, myself and @yukoosawa chatting today and here's the plan:

1. repositioning the close icon - if you can do this, great, but timebox the effort from now to a couple of hours max, if you can't figure out a nice way to do it by then, then we don't do the repositioning.

2. remove the close button below the image. the issue we have is that the background scroll causes it to appear where it shouldn't (ie products with less content where it shouldn't display, first open it doesn't display, second open it does). We don't need this button for now, we'll add it once the background scroll bug is fixed.

These 2 things unblock this PR, so full steam ahead.

luisramos0 commented:
I just rebased to resolve conflicts.

luisramos0 commented:
ok, thanks for the update @daniellemoorhead 
re point 1, I have already spent quite some time moving the button around without success. Maybe this is easy for someone else (see my [comment above](https://github.com/openfoodfoundation/openfoodnetwork/pull/5263#issuecomment-628104132) for details)
re point 2, I have removed the close button from this PR in a new commit.

You wrote "These 2 things unblock this PR" but this issue is still marked as blocked by 5397:
![image](https://user-images.githubusercontent.com/1640378/82045368-e5102700-96a6-11ea-9f45-4e033cc134aa.png)
Can you please clarify?

If ok to proceed, I'll move to code review so at least one reviewer approves my extra commit (I deleted the close button as a single extra commit so that we can revert it once we want to add the button again after #5397).

daniellemoorhead commented:
I forgot about that decency blocker, only took the label off. You can remove that dependency and movie it across to test. 

> On 15 May 2020, at 9:27 pm, Luis Ramos <notifications@github.com> wrote:
> 
> ﻿
> ok, thanks for the update @daniellemoorhead
> re point 1, I have already spent quite some time moving the button around without success. Maybe this is easy for someone else (see my comment above for details)
> re point 2, I have removed the close button from this PR in a new commit.
> 
> You wrote "These 2 things unblock this PR" but this issue is still marked as blocked by 5397:
> 
> Can you please clarify?
> 
> If ok to proceed, I'll move to code review so at least one reviewer approves my extra commit (I deleted the close button as a single extra commit so that we can revert it once we want to add the button again after #5397).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

luisramos0 commented:
ok, thanks.

luisramos0 commented:
ping @Matt-Yorkley or @sauloperez could you validate this last commit please? Thanks!

daniellemoorhead commented:
### Testing notes

Devices:
- iPhone SE + Firefox
- iPhone 8 + Safari
- iPad mini 4 + Safari
- Windows 10 + Chrome
- Mac OSX Mojave + Chrome
- Galaxy 8 + Chrome
- Pixel 4 + Chrome

Results:
2 issues remain:
- [ ] Not all images are centre aligned to the modal so there is different spacing between each side. [Screenshot product is not aligned correctly](https://staging.openfoodnetwork.org.au/thornbury-community-food-hub/shop), but then [this other product is](https://staging.openfoodnetwork.org.au/beetroot-bonanza/shop#/shop). 
- [ ] Product name goes underneath the close icon, so you can't read it.

<img width="377" alt="Screen Shot 2020-05-19 at 7 29 23 pm" src="https://user-images.githubusercontent.com/10949052/82310395-add3aa00-9a07-11ea-9ea8-fab0716fae93.png">

Note that I checked if that image had a larger white area to the right, it didn't.
<img width="760" alt="Screen Shot 2020-05-19 at 7 37 04 pm" src="https://user-images.githubusercontent.com/10949052/82310839-3baf9500-9a08-11ea-9b19-4c0822a5ae91.png">

sauloperez commented:
@luisramos0 I'm afraid this close button will require tweaking the modal layout a bit. Apparently it uses absolute positioning but its closest relatively positioned parent doesn't let us move the button outside the modal. A negative margin doesn't work either.

Do we still want to position it a bit off the modal though? I don't know where the discussion ended sorry.

I'll spend the whole day on OFN tomorrow so I can lend you a hand if we want.

luisramos0 commented:
Hi @daniellemoorhead 
I fixed the problem of the close button overlapping with the title in this last commit.

I didnt notice the tablet view needed to be changed. It's now keeping the image on the right :+1:

I dont think these need code review so you can stage and test again :+1: 

The images sometimes get centered sometimes they don't. That's how they are in prod, I thought the designs were not requiring that they are always centered. Can we create a separate issue for that?
![image](https://user-images.githubusercontent.com/1640378/82835626-8162d680-9ebc-11ea-886b-7ac0d0797508.png)

daniellemoorhead commented:
>The images sometimes get centered sometimes they don't. That's how they are in prod, I thought the designs were not requiring that they are always centered. Can we create a separate issue for that?

Is there a reason it needs to be a separate issue and can't be done in this PR @luisramos0? It was always a requirement of the redesign, so not sure why it has to have a new story made. 

I'll test the other stuff now!

luisramos0 commented:
We should aim to get PRs stream lined as quickly as possible through the pipe, PRs with lots of back and forth like this one introduce unnecessary risk to our process. I'd say the question is: is there a reason to make it part of this PR? Can we fix it in a separate PR?

luisramos0 commented:
The context is important here: there are 19 commits in this PR with quite a few required points addressed successfully and the PR has been created 1 month ago. I think these are all good reasons to merge asap.

daniellemoorhead commented:
And yet, from a product perspective it's less time for me to have it done via this story rather than me having to create another story. Thus why I ask the question. 

So, will centring this take a lot of time and therefore not a quick change, or can you make commit number 20 with a quick turnaround?

daniellemoorhead commented:
Perhaps the happy medium is for you to create another PR, attached to this issue, and we can merge the other PR? That means I don't have to create another issue (given that requirement is already in this one), and you get to close this PR and put it in another. 

Does that work for you?

luisramos0 commented:
Sorry @daniellemoorhead I don't think I was clear and, anyway, my previous comments were based on my erroneous conclusion that centering the image would not be easy. I wonder how did I get to that conclusion!
After your comments I just gave it a try and saw how easy it was :man_shrugging: 
So, this is back to Code Review with a new commit that centers the images!

My original point is still valid though: it's a bit painful to watch these PRs going back and forth In Dev, Code Review and Test Ready. I am happy this is not our normal dev process. Anyway, we should be done with this one now!

luisramos0 commented:
test ready :crossed_fingers: :crossed_fingers: :crossed_fingers:

daniellemoorhead commented:
Retested, images are now centre aligned, tested on samsung 8, iphone 8, ipad 6th gen.

Good to go 🎉

luisramos0 commented:
nice :heart_eyes:
thanks @daniellemoorhead :-)
